### ISSUE TITLE
Find the best tree in the random forest

### ISSUE THREAD

**Issue by swager** [swager/grf#281]
The idea behind a random forest is that, by averaging predictions from many trees, we can do better than any single tree alone could. However, in some instances (e.g., for visualization), it would be helpful to have a single tree that represents the signal as well as it can. Two baseline approaches to this problem would be the following:

- Just grab a single tree out of the forest (e.g., the first one). This is easy, but the resulting tree may not be particularly good: In a random forest, we on purpose make the trees a little bit noisy to explore a larger fraction of the feature space, but this means that single trees in the forest may be less accurate.
- Fit a new ad hoc tree, e.g., via CART with pruning, or mixed integer programming as recommended by Bertsimas and Dunn (2017). The former, however, is still a greedy algorithm (so may not be optimal), whereas approaches based on MIP require specialized solvers and only scale to thousands of observations.

It may be worth exploring a 3rd idea that seeks to leverage the exploration done by a forest to find a good tree. Essentially, with a random forests, we already have 2000 trees that explore various ways of representing the data. If we pick the best of those 2000, we might recover a tree that is better than just running CART once, but more practical on large datasets than solving a MIP.

The proposed approach is as follows. Assuming honesty, each tree divides the training examples into a splitting set (S), and honest estimation set (H), and an out-of-bag set (O). Everything that's specific to a single tree b = 1, ..., B is subscripted with b.

1. Train a forest, as is already done. Denote the m-th leaf of the b-th tree by L_bm. Recall that the set {L_b.} is chosen only using S_b.
2. Compute the ''raw'' loss R_bm of node L_bm, using cross-validated error on the honest estimation set H_b; see details below.
3. Compute the ''pruned'' loss P_bm of a node recursively. If L_bm is a leaf, then simply P_bm = R_bm. If L_bm is not a leaf, then P_bm also depends on the purities of the two children of L_bm, called here L_bm' and L_bm'': P_bm = min{R_bm, c + P_bm' + P_bm''}. This definition reflects a choice to prune: Either we keep the split below L_bm, in which case we pay the loss of the children; or we prune it away, and just pay the loss of the node itself. Here, c > 0 is a tuning parameter that encourages simplicity by telling the tree to avoid splits in case of doubts. 
4. Select as b* the tree with the smallest root-node impurity L_b1 (note that this is a pruned version of the b*-th tree in the original forest, as reflected by the loss criterion in step 3).
5. Use the out-of-bag set O to estimate parameter values + confidence intervals in each node of the selected tree.

In step 2, we should use a cross-validated loss criterion to properly account for sampling variance in the leaf. In the case of regression, this is simple:
```
R_bm = \sum_{i in H_b : X_i in L_bm} (Yi - avg{Yj : j in H_b, X_j in L_bm, j!= i})^2
= N_bm^2 / (N_bm - 1)^2 \sum_{i in H_b : X_i in L_bm} (Yi - \hat{Y}_bm)^2 
```
where N_bm is the number of observation in H_b falling into N_bm, and \hat{Y}_bm =  avg{Yj : j in H_b, X_j in L_bm} is the parameter estimate in L_bm computed using H_b.

In the case of a causal forest, we should compute loss based on the R-learning criteria, as we currently do for tuning. A similar expression for R_bm is available via the ``shortcut-formula'' for ordinary cross validation in linear models (where the residuals a inflated by 1/(1 - self_influence)^2).

There are some engineering choices that need to be made in implementing this; in particular:

- We could perform step 2 on the fly while training the forest, or as a post-processing step on the trained forest. The former seems simpler, but has the downside of making the main training procedure slower. I guess this depends on how long step 2 takes, and how many users will want to re-run steps 3 and 4 with different value of c (and whether it's OK to sometimes re-do the computations in 2 many times).
- Do we need to do step 3 in c++, or would an R implementation be fast enough?

**Comment by swager** [comment]
@clu0 is working on this now.

**Comment by clu0** [comment]
I have completed a draft for this function for the regression forest, which can be found here:

https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b

The code takes about 1.5 minutes to run for 2000 trees, so I don't know if it is worth optimizing.

**Comment by msegar** [comment]
> I have completed a draft for this function for the regression forest, which can be found here:
> 
> https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b

I have updated clu0's 'find best tree' script to work with GRF version 0.10.3
https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5

**Comment by ginward** [comment]
> > I have completed a draft for this function for the regression forest, which can be found here:
> > https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b
> 
> I have updated clu0's 'find best tree' script to work with GRF version 0.10.3
> https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5

Since this tree is grown with a random-subsample of the data, can I retrieve the characteristics of the leaf samples (like average of the covariates) of the best pruned tree? @swager

**Comment by ginward** [comment]
Do you imagine that we can have a function that directly export the node characteristics for the best pruned tree? For example, the average age, gender, etc.

**Comment by ginward** [comment]
@swager What value do you recommend setting the cost parameter here to? 
https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5#file-best_tree-r-L129

I usually set it to 1. Do you think it is fine that way?

**Comment by swager** [comment]
That's not a parameter I have much experience in tuning. In addition to making this faster, figuring out how to think about defaults in a robust way is another thing we'd need to do before adding this to the package.

**Comment by ginward** [comment]
> > I have completed a draft for this function for the regression forest, which can be found here:
> > https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b
> 
> I have updated clu0's 'find best tree' script to work with GRF version 0.10.3
> https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5

@msegar @clu0  I noticed that the [`type `argument](https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5#file-best_tree-r-L129) only takes `"causal"` and `"regression"`. Does the `causal` argument work for instrumental forest as well?

**Comment by ginward** [comment]
> > I have completed a draft for this function for the regression forest, which can be found here:
> > https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b
> 
> I have updated clu0's 'find best tree' script to work with GRF version 0.10.3
> https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5

Does this function work for instrumental forest as well?

**Comment by ginward** [comment]
> > > I have completed a draft for this function for the regression forest, which can be found here:
> > > https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b
> > 
> > 
> > I have updated clu0's 'find best tree' script to work with GRF version 0.10.3
> > https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5
> 
> Does this function work for instrumental forest as well?

I guess this function (at this point) should not be applied to instrumental forest, as the R-learner loss function in Nie and Wager (2019) didn't consider instrumental variables and assume strict unconfoundedness.

**Comment by ginward** [comment]
> > > > I have completed a draft for this function for the regression forest, which can be found here:
> > > > https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b
> > > 
> > > 
> > > I have updated clu0's 'find best tree' script to work with GRF version 0.10.3
> > > https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5
> > 
> > 
> > Does this function work for instrumental forest as well?
> 
> I guess this function (at this point) should not be applied to instrumental forest, as the R-learning loss function in [Nie and Wager (2019)](https://arxiv.org/abs/1712.04912) didn't consider instrumental variables and assume strict unconfoundedness.

@swager @clu0 @msegar  On line 28, why do you have: 

`unscaled_spread <- sum((Y[samples] - M.hat[samples] - (W.orig[samples] - W.hat[samples])*mean(Tau.hat[samples]))^2)
`
Instead of: 

`unscaled_spread <- sum((Y[samples] - M.hat[samples] - (W.orig[samples] - W.hat[samples])*Tau.hat[samples])^2)`

I read the Nie and Wager (2019) paper and it seems that the R-learner loss is: 

![Screenshot 2019-11-15 at 14 44 08](https://user-images.githubusercontent.com/6193977/68951625-63eadc80-07b6-11ea-8503-fbb36eeb0009.png)

There doesn't seem to be a mean of the estimated treatment effect here?

**Comment by ginward** [comment]
> > > > I have completed a draft for this function for the regression forest, which can be found here:
> > > > https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b
> > > 
> > > 
> > > I have updated clu0's 'find best tree' script to work with GRF version 0.10.3
> > > https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5
> > 
> > 
> > Does this function work for instrumental forest as well?
> 
> I guess this function (at this point) should not be applied to instrumental forest, as the R-learning loss function in [Nie and Wager (2019)](https://arxiv.org/abs/1712.04912) didn't consider instrumental variables and assume strict unconfoundedness.

I added instrumental forest support to the preliminary best tree algorithm. 

https://gist.github.com/ginward/451043145ef914f57af5a7272cf02489

**Comment by tianshengwang** [comment]
> > I have completed a draft for this function for the regression forest, which can be found here:
> > https://gist.github.com/clu0/47f86cd2d34216febfee4bafe2dbe46b
> 
> I have updated clu0's 'find best tree' script to work with GRF version 0.10.3
> https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5

 @clu0 @msegar @swager  I'm not sure how to implement the best_tree script: when running the find_best_tree function using the example code, the output is a list of tree index (100, 200, ..., 2000) rather than the single "best tree" index, and each tree in the output looks different (the 100th and 2000th tree plot shown below):
```
library("grf")
#' # Train a regression forest
n <- 50; p <- 10
X <- matrix(rnorm(n*p), n, p)
Y <- X[,1] * rnorm(n)
forest <- grf::regression_forest(X,Y)

# Find the best tree
best_tree_info <- find_best_tree(forest, "regression")

library("DiagrammeR")
tree.plot100 = plot(get_tree(forest, 100))
tree.plot100

tree.plot2000 = plot(get_tree(forest, 2000))
tree.plot2000
```
the output is:
tree: 100 
tree: 200 
tree: 300 
tree: 400 
tree: 500 
tree: 600 
tree: 700 
tree: 800 
tree: 900 
tree: 1000 
tree: 1100 
tree: 1200 
tree: 1300 
tree: 1400 
tree: 1500 
tree: 1600 
tree: 1700 
tree: 1800 
tree: 1900 
tree: 2000 

![Tree_100](https://user-images.githubusercontent.com/40049497/72213281-1b9c2200-34ba-11ea-8ab1-d6eb861ab5f8.png)
![Tree_2000](https://user-images.githubusercontent.com/40049497/72213282-21920300-34ba-11ea-99dc-83207df6bf31.png)

**Comment by erikcs** [comment]
@tianshengwang looking at the code it seems

```R
best_tree_info$best_tree
```
should be the best tree

**Comment by tianshengwang** [comment]
> @tianshengwang looking at the code it seems
> 
> ```r
> best_tree_info$best_tree
> ```
> 
> should be the best tree

Thanks!

**Comment by zhong1969** [comment]
Sorry to disturb you. I cannot open :

https://gist.github.com/ginward/451043145ef914f57af5a7272cf02489

I want to visualize a best  instrumental forest.

Could you please check? Thanks!

**Comment by tianshengwang** [comment]
> Sorry to disturb you. I cannot open :
> 
> https://gist.github.com/ginward/451043145ef914f57af5a7272cf02489
> 
> I want to visualize a best instrumental forest.
> 
> Could you please check? Thanks!

thanks, @ginward explained here: https://github.com/grf-labs/grf/issues/558#issuecomment-576007340
It worked once only for me, but I'm still not sure how to visualize a best instrumental forest either.

**Comment by ginward** [comment]
@tianshengwang what didn't work for you?

**Comment by tianshengwang** [comment]
@erikcs @msegar @ginward  Is there any way to extract the split_variable and split_value from get_tree(cf, best_tree_info$best_tree)? Because I am doing a simulation study, hope to automatically get such info from training data and then apply it (i.e. split population) in testing data. E.g. for the output below, how can I extract the split_variable and it's splitting order and split_value?
`> get_tree(cf, best_tree_info$best_tree)`
GRF tree object 
Number of training samples: 12500 
Variable splits: 
(1) split_variable: w3  split_value: 0 
  (2) split_variable: w2  split_value: 0.0864782 
    (4) * num_samples: 1666  avg_Y: -3.4 avg_W: 0.52 
    (5) * num_samples: 1412  avg_Y: -3.93 avg_W: 0.37 
  (3) split_variable: w1  split_value: 0 
    (6) * num_samples: 1644  avg_Y: -4.68 avg_W: 0.51 
    (7) * num_samples: 1528  avg_Y: -5.1 avg_W: 0.63 

Thanks!

**Comment by erikcs** [comment]
@tianshengwang what exactly is the output you are after?

**Comment by tianshengwang** [comment]
> @tianshengwang what exactly is the output you are after?
I need the following info (w1,w3 are binary variables, thus cutoff value=0):
the first split_variable (w3) and it's split_value (0)
the second split_variable (w2) and it's split_value (0.0864782)
the third split_variable (w1) and it's split_value (0)

**Comment by tianshengwang** [comment]
> @tianshengwang what exactly is the output you are after?

never mind, I figured it out, thanks!

**Comment by erikcs** [comment]
@tianshengwang could you share an example of what you wanted? Maybe we can add functionality to retrieve it in #661. Thanks!

**Comment by tianshengwang** [comment]
@erikcs Thanks for mentioning  [#661 ](https://github.com/grf-labs/grf/issues/661). I want an automatical procedure that could identify split_variable and split_value in the training forest so that I could use such leaf decision (split_variable and split_value) to stratify test data population (which is similar to @yifan-cui 's goal in [#661 ](https://github.com/grf-labs/grf/issues/661) "Goal: For an individual tree and a given test point, it would be nice if one can identify observations associated with the leaf that includes the test point"). So far we've only figured out how to extracting such info for a 4-leaf regular binary tree shown below. 
![image](https://user-images.githubusercontent.com/40049497/82757213-e3341b00-9dac-11ea-90c5-65164bfd0763.png)
node_information_extract <- function(number_of_nodes = 3) {
  node = rep(NA, number_of_nodes)
  node_split_variable = rep(NA, number_of_nodes)
  node_split_value = rep(NA, number_of_nodes)
  for (i in 1 : number_of_nodes) {
    node[i] <- paste('node', i, sep = '-')
    node_split_variable[i] <- besttree$columns[besttree$nodes[[i]]$split_variable]
    node_split_value[i] <- besttree$nodes[[i]]$split_value
  }
  node_information <- data.frame(node, 
                                 node_split_variable, 
                                 node_split_value, stringsAsFactors = F)
  return(node_information)
}
But there are 4 other scenarios even for a simple 4-leaf trees, e.g. left skewed shown below. Haven't figured out other scenarios yet. Adding such functionality to extract splitting variable/value according for a simple tree (with a small number of leaf) will be very helpful.
![image](https://user-images.githubusercontent.com/40049497/82757322-ab79a300-9dad-11ea-8bdb-14e25fc1dd0a.png).

**Comment by hakianmehr** [comment]
@clu0 @msegar @swager my final best tree is different to the variable importance? what is the problem? to my knowledge, the tree is not consistent with the importance list. right?
![image](https://user-images.githubusercontent.com/65143373/89591451-33514400-d818-11ea-8de9-19e195fd491f.png)

Variable imporatnce
--
1 | ualb | 0.10363
2 | alt | 0.102367
3 | bmiinit | 0.090782
4 | potassium.x | 0.08498
5 | fpg | 0.079913
6 | ldl | 0.053235
7 | gfr | 0.047391
8 | ucreat | 0.046057
9 | trig | 0.045761
10 | hdl | 0.04335
11 | cpk | 0.041868
12 | hgi | 0.040572
13 | sbp | 0.036591
14 | dbp | 0.035873
15 | chol | 0.031624
16 | x2mi | 0.02752
17 | yrsdiab | 0.025696
18 | hba1cinit | 0.025285
19 | cabg | 0.019307
20 | vldl | 0.012395
21 | ksparing | 0.003979
22 | loop | 0.001826

**Comment by tianshengwang** [comment]
> @clu0 @msegar @swager my final best tree is different to the variable importance? what is the problem? to my knowledge, the tree is not consistent with the importance list. right?
> ![image](https://user-images.githubusercontent.com/65143373/89591451-33514400-d818-11ea-8de9-19e195fd491f.png)
> 
> ## Variable imporatnce
> 1 | ualb | 0.10363
> 2 | alt | 0.102367
> 3 | bmiinit | 0.090782
> 4 | potassium.x | 0.08498
> 5 | fpg | 0.079913
> 6 | ldl | 0.053235
> 7 | gfr | 0.047391
> 8 | ucreat | 0.046057
> 9 | trig | 0.045761
> 10 | hdl | 0.04335
> 11 | cpk | 0.041868
> 12 | hgi | 0.040572
> 13 | sbp | 0.036591
> 14 | dbp | 0.035873
> 15 | chol | 0.031624
> 16 | x2mi | 0.02752
> 17 | yrsdiab | 0.025696
> 18 | hba1cinit | 0.025285
> 19 | cabg | 0.019307
> 20 | vldl | 0.012395
> 21 | ksparing | 0.003979
> 22 | loop | 0.001826



> @clu0 @msegar @swager my final best tree is different to the variable importance? what is the problem? to my knowledge, the tree is not consistent with the importance list. right?
> ![image](https://user-images.githubusercontent.com/65143373/89591451-33514400-d818-11ea-8de9-19e195fd491f.png)
> 
> ## Variable imporatnce
> 1 | ualb | 0.10363
> 2 | alt | 0.102367
> 3 | bmiinit | 0.090782
> 4 | potassium.x | 0.08498
> 5 | fpg | 0.079913
> 6 | ldl | 0.053235
> 7 | gfr | 0.047391
> 8 | ucreat | 0.046057
> 9 | trig | 0.045761
> 10 | hdl | 0.04335
> 11 | cpk | 0.041868
> 12 | hgi | 0.040572
> 13 | sbp | 0.036591
> 14 | dbp | 0.035873
> 15 | chol | 0.031624
> 16 | x2mi | 0.02752
> 17 | yrsdiab | 0.025696
> 18 | hba1cinit | 0.025285
> 19 | cabg | 0.019307
> 20 | vldl | 0.012395
> 21 | ksparing | 0.003979
> 22 | loop | 0.001826

probalby because sample size too small, split due to chance.

**Comment by hakianmehr** [comment]
Thank you very much for your response.

Is there any way to deal with a small dataset? I just used the below
specification for the "casual_forest" but still, I couldn't get the stable
tree.

c.forest <- causal_forest(X1, Y, W,
                          honesty = FALSE,
                          honesty.prune.leaves=FALSE,
                          sample.fraction=0.90,
                          ci.group.size=1)
best_tree_info <- find_best_tree(c.forest,"causal")
best_tree_info $best_tree
index<-best_tree_info $best_tree
plot(tree <- get_tree(c.forest, index))

**Comment by swager** [comment]
There is no guarantee that the "best" tree would split along the "most important" variables as given by our variable importance measure. The variable importance measure captures which variables are split on the most on average -- but the best tree doesn't need to split first on the variable that is most commonly split on.

@hakianmehr if you have further questions, please open a new issue. This issue should only be about implementing the `find_best_tree` feature.

**Comment by tianshengwang** [comment]
@swager I don't understand your explanation on "**N_bm**" for the raw loss R_bm formula, you mentioned "where **N_bm** is the number of observation in H_b falling into **N_bm**", do you actually mean "where N_bm is the number of observation in H_b falling into **L_bm** (the m-th leaf of the b-th tree)"?

**Comment by liuweie** [comment]
Is the  'find_best_tree'  function removed? I can't find this function

**Comment by tianshengwang** [comment]
> Is the 'find_best_tree' function removed? I can't find this function

https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5

**Comment by tianshengwang** [comment]
> Is the 'find_best_tree' function removed? I can't find this function

https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5

**Comment by liuweie** [comment]
> > Is the 'find_best_tree' function removed? I can't find this function
> 
> https://gist.github.com/msegar/c15af2bfc067e5319eecfd67d785fde5

Thanks!
